ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -384,58 +385,66 @@ KeyField findKeyField(final List<SelectExpression> selectExpressions) {
   public SchemaKStream<Struct> selectKey(
-      final ColumnRef columnRef,
+      final Expression keyExpression,
       final QueryContext.Stacker contextStacker
   ) {
     if (keyFormat.isWindowed()) {
       throw new UnsupportedOperationException("Can not selectKey of windowed stream");
     }
 
-    final Optional<Column> existingKey = keyField.resolve(getSchema());
-
-    final Column proposedKey = getSchema().findValueColumn(columnRef)
-        .orElseThrow(IllegalArgumentException::new);
-
-    final KeyField resultantKeyField = isRowKey(columnRef)
-        ? keyField
-        : KeyField.of(columnRef);
-
-    final boolean namesMatch = existingKey
-        .map(kf -> kf.ref().equals(proposedKey.ref()))
-        .orElse(false);
-
-    if (namesMatch || isRowKey(proposedKey.ref())) {
-      return (SchemaKStream<Struct>) new SchemaKStream<>(
-          sourceStep,
-          keyFormat,
-          resultantKeyField,
-          sourceSchemaKStreams,
-          type,
-          ksqlConfig,
-          functionRegistry
-      );
+    if (!needsRepartition(keyExpression)) {
+      return (SchemaKStream<Struct>) this;
     }
 
-    final KeyField newKeyField = getSchema().isMetaColumn(columnRef.name())
-        ? KeyField.none()
-        : resultantKeyField;
-
     final StreamSelectKey step = ExecutionStepFactory.streamSelectKey(
         contextStacker,
         sourceStep,
-        columnRef
+        keyExpression
     );
+
     return new SchemaKStream<>(
         step,
         keyFormat,
-        newKeyField,
+        getNewKeyField(keyExpression),
         Collections.singletonList(this),
         Type.REKEY,
         ksqlConfig,
         functionRegistry
     );
   }
 
+  private KeyField getNewKeyField(final Expression expression) {
+    if (!(expression instanceof ColumnReferenceExp)) {
+      return KeyField.none();
+    }
+
+    final ColumnRef columnRef = ((ColumnReferenceExp) expression).getReference();
+    final KeyField newKeyField = isRowKey(columnRef) ? keyField : KeyField.of(columnRef);
+    return getSchema().isMetaColumn(columnRef.name()) ? KeyField.none() : newKeyField;
+  }
+
+  private boolean needsRepartition(final Expression expression) {
+    if (!(expression instanceof ColumnReferenceExp)) {
+      return true;
+    }
+
+    final ColumnRef columnRef = ((ColumnReferenceExp) expression).getReference();
+    final Optional<Column> existingKey = keyField.resolve(getSchema());
+
+    final Column proposedKey = getSchema()
+        .findValueColumn(columnRef)
+        .orElseThrow(() -> new KsqlException("Invalid identifier for PARTITION BY clause: '"
+            + columnRef.name().toString(FormatOptions.noEscape()) + "' Only columns from the "
+            + "source schema can be referenced in the PARTITION BY clause."));
+
+
+    final boolean namesMatch = existingKey
+        .map(kf -> kf.ref().equals(proposedKey.ref()))
+        .orElse(false);
+
+    return !namesMatch && !isRowKey(columnRef);
```

Comment:
cc @big-andy-coates - I think @rodesai brings up a good point here, though I think it's an existing bug. I think we should fix it in the short-term by re-introducing the "unnecessary" repartition step in the case of `isRowKey` but before I do that I wanted to run it past you since you implemented the original optimization (#2735).

Reply:
I'd suggest adding this as a QTT test to ensure we're handling this correctly.  Do what you need to do to get this functionally working. 

Raise a github issue to track the outstanding piece.

I think we can clean this up once we have cleaner code around the handling of schemas and duplicating ROWKEY and ROWTIME into the value schema.  

Specifically, once we clean up / remove the use of source-aliases on schema fields and have arbitrary key column names, then I think this will be easy to solve because the `ROWKEY` name won't be ambiguous.

Reply:
https://github.com/confluentinc/ksql/pull/4032/commits/f8bfefa6d2d3d0f0a2d6281764128790cc2cf2ac - I will fix this in a future PR

Reply:
https://github.com/confluentinc/ksql/issues/4053 - I will fix this in a follow-up PR